### DISCUSSION TITLE
if skywalking supports analysis or statistics via a parameter in the http url as an endpoint.

### DISCUSSION THREAD

**Discussion by lijiawen** [apache/skywalking#8850] [Q&A]
Good Afternoon, I would like to ask if skywalking supports analysis or statistics via a parameter in the http url as an endpoint.

Our usage scenario is like this, 
we only provide one url path to our users, all user requests are like this: `https://ourdomain.com/gateway?params=values&method=aaa`, the specific business forwarding rules depend on one of the `method` request parameters, 
we want to monitor the average response time of different `method`, timeout and so on through skywalking, please ask if it can be achieved through `agent.config` configuration file (or can be modified through the source code to achieve similar functions), thanks.

**Comment by wu-sheng** [comment]
If you are using different Spring methods then yes. Otherwise, I think we don't have out-of-box option. But this is not hard to add.

**Reply by lijiawen** [reply]
> If you are using different Spring methods then yes. Otherwise, I think we don't have out-of-box option. But this is not hard to add.

Thanks for very timely response, it's really helpful. Unfortunately we only use one Spring methods, and redirect requests to different services by the  `method` parameter.
So does it mean We will consider adding a new configuration to support this feature.

**Reply by wu-sheng** [reply]
Yes, because you already have this as one endpoint from Spring level, you disabled all relative features from official plugins.

**Reply by wu-sheng** [reply]
Consider taking a look at the logic endpoint doc, I think that is the easiest way to get metrics of your logic methods directly.

**Reply by lijiawen** [reply]
Thanks for the reply, I'll look into the usage doc.

**Reply by lijiawen** [reply]
@wu-sheng sorry to ask again, but I don't quite understand something in the documentation. I don't need to develop a new plugin separately?  do I need to write a new configuration in `agent.config`? I would appreciate if there are some foolproof examples, or a little explanation. ~_~

**Reply by wu-sheng** [reply]
Are you reading this? https://skywalking.apache.org/docs/main/v9.0.0/en/guides/backend-oal-scripts/#logic-endpoint

If so, you need our agent toolkit(a very light jar lib), and use its APIs to tag accordingly.

**Reply by lijiawen** [reply]
Okay, just read the document is not this version, this document is very detailed, thanks for providing.

**Reply by lijiawen** [reply]
@wu-sheng Sorry to interrupt, I would like to ask the documentation mentioned a json configuration,like this:{{
  "name": "GraphQL-service",
  "latency": 100,
  "status": true
}}, is it to write to the plug-in somewhere, or do i need to achieve some api interface by myself, the specific implementation of which api interface can help to explain?

**Reply by lijiawen** [reply]
i guess maybe i should use an annotation like @Tag or @Trace on my monitor method, and skywalking can auto collect this metrics.

**Reply by wu-sheng** [reply]
Both work, your choice.

**Reply by lijiawen** [reply]
i found this doc from google: https://blog.csdn.net/a17816876003/article/details/113702511 , maybe i can work in this way?

**Reply by wu-sheng** [reply]
Take a look at official documents, I don't comment on others' blogs, that is unfair.

**Reply by lijiawen** [reply]
OK, understand.

1. I would like to ask, where is the json configuration mentioned in the documentation placed, is it in a custom plugin? Do I need to understand the implementation principle of skywalking-java-agent first? I downloaded the corresponding source code `skywalking-java-agent` and saw that there are some scenarios files with `{"logic-span":true}` configuration, but these seem to be some test cases, and I didn't find out how to actually use this configuration through the source code, where should I put them? If this option needs to understand the source code implementation principle first, and at the same time need to implement custom inserts, I think the second option may be used.

2. Option 2, introduce maven dependency `apm-toolkit-trace` in the project, and then configure Skywalking's Trace annotation in the corresponding business method, so that you can do it without doing anything else, and at the same time you can check these endpoints in the UI of skywalking.

thanks.

**Reply by wu-sheng** [reply]
> where is the json configuration mentioned in the documentation placed, is it in a custom plugin?

It is a tag, you could use it any way you like to add one. You don't need to know how other plugins work. You add a tag to your span or current span in the context with the proper name, then the backend will identify this automatically. Writing a new plugin is an option for sure, if you don't want to add any lib or API in your codes.

Option-2 works too.

**Reply by lijiawen** [reply]
Okay, I'll give it a try, thanks for clearing up the confusion.

**Comment by lijiawen** [comment]
![image](https://user-images.githubusercontent.com/3766355/162969472-e3748290-4bd6-4ab4-935e-582b6598fa79.png)

**Comment by lijiawen** [comment]
@wu-sheng , sorry to interruption again, after adding Trace annotation, I can see the `operationName` set in Trace is OK, and there are 2 questions, 1. how to support querying custom endpoint(in example is `place_orderSTANDARD`) at the first red box in the above img, 2. how to set the type and component of custom trace, now the default is unknown,unknown. thanks.

**Reply by wu-sheng** [reply]
You already have an endpoint name for the search, so, I assume this would not work, unless you use another tag, and set it as a searchable tag(backend configuration).

The logic endpoint is for statistics, on the dashboard, not for trace queries.

Type and component are not available for logic endpoint, as it is not an RPC component at all.

**Reply by lijiawen** [reply]
Ok, got it, maybe I should implement a custom plugin.

**Comment by lijiawen** [comment]
[ERROR] /Users/tiger/Work/Project/Tiger/Github/skywalking-java/apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/command/TraceIgnoreCommand.java:[21,51] 程序包org.apacskywalking.apm.network.common.v3不存在

@wu-sheng , There is a compile error when compiling skywalking-java-agent for 8.2.0 and 8.10.0-SNAPSHOT versions, how can I fix it?

**Reply by lijiawen** [reply]
execute command in project home directory:
 mvn clean compile  -Dmaven.test.skip=true

**Reply by wu-sheng** [reply]
Read compiling doc. As always, find the doc because take actions, it could save a lot of time.

**Reply by lijiawen** [reply]
Good advice, thanks.
i follow this doc: https://github.com/apache/skywalking/blob/master/docs/en/guides/How-to-build.md
download source code from release version 8.2.0
execute: ./mvnw clean package -Dmaven.test.skip

and i found in apm-protocol > apm-network maven module , the `proto` directory is empty, any relations with this? 
after execute ./mvnw , seems the same error, any ideas?

**Reply by lijiawen** [reply]
maybe this is the reason, hope to update doc:
If you download in zip ,all proto files do not exist ,you should download proto files again and put them into /apm-protocol/apm-network/src/main/proto

**Reply by lujiajing1126** [reply]
> Good advice, thanks. i follow this doc: https://github.com/apache/skywalking/blob/master/docs/en/guides/How-to-build.md download source code from release version 8.2.0 execute: ./mvnw clean package -Dmaven.test.skip

You just missed the step before `./mvnw clean package -Dmaven.test.skip`. Please follow the guidance step by step.

**Reply by lijiawen** [reply]
@lujiajing1126 , ok, within the lastest version , it seems OK, but if i download history release version from official git repository(zip files format), there will be probliems. thanks any way.

**Reply by wu-sheng** [reply]
>  from official git repository(zip files format)

Never do this. We never recommend you to do so. We say `git clone`, that has a reason.

**Reply by wu-sheng** [reply]
If you want to download a source tar bar, download from our website. These two are all listed on the doc.

Remember, docs are there for a reason.

**Reply by wu-sheng** [reply]
> maybe this is the reason, hope to update doc: If you download in zip ,all proto files do not exist ,you should download proto files again and put them into /apm-protocol/apm-network/src/main/proto

Update what? https://skywalking.apache.org/docs/main/v9.0.0/en/guides/how-to-build/ The doc never says so.

**Comment by lijiawen** [comment]
![image](https://user-images.githubusercontent.com/3766355/163147610-47395ba2-299e-409c-822d-fec1e807ecac.png)

**Comment by lijiawen** [comment]
@lujiajing1126 , everything seems OK except this stage.

**Reply by lujiajing1126** [reply]
The problem here is exactly the same as the last one, submodules are not ready.
As @wu-sheng commented, you must not download a tarball from GitHub for legacy versions.

Please follow the How-To-Build document: use `git clone` and then `git checkout` the version you want.

**Reply by lijiawen** [reply]
@lujiajing1126 , but this version is download from latest github repository
![image](https://user-images.githubusercontent.com/3766355/163151339-6f2e7f1d-05d4-4119-befd-637d4d8a0ced.png)

**Comment by lijiawen** [comment]
@wu-sheng , @lujiajing1126 , Downloading historical versions has this requirement, and needs to match the version of skywalking deployed on our server to prevent compatibility issues.

**Reply by wu-sheng** [reply]
All released versions have been verified. I am not sure what it is. If you want to download the full source code package, and you have issues on GitHub, download it from skywalking.apache.org. This is why the Apache Software Foundation always hosts the release source codes on their own servers, which have global mirrors including in China.

**Reply by lijiawen** [reply]
I have not tried to write the 8.11.0-SNAPSHOT version of the skywalking plugin, will it conflict with the 8.2.0 version of the server

**Reply by lijiawen** [reply]
![image](https://user-images.githubusercontent.com/3766355/163153045-db8803e6-1b9b-4533-8b55-8b9e42121a04.png)

**Reply by wu-sheng** [reply]
Compatibility is the server moving ahead of an agent, not in reverse. No agent could predict how server-side would change.

**Reply by lijiawen** [reply]
where can i found v8.2.0 from apache, this is another question?

**Reply by lijiawen** [reply]
> 
OK, i will try it, writing pulgin in lastest version agent, thanks so much.

**Reply by wu-sheng** [reply]
> where can i found v8.2.0 from apache, this is another question?

All versions are available here https://archive.apache.org/dist/skywalking/. But no mirror for versions not listed. So, you need to download from ASF US server, but you should be good.

**Reply by lijiawen** [reply]
@wu-sheng , logic endpoint feature added from v8.9 , does it mean we should upgrade  our OAP server version to  v.8.9(now is v8.2), Otherwise it will not be compatible.

**Reply by wu-sheng** [reply]
I am not sure about 8.2, but it was not added in 8.9.

It is at least existing in 8.4.0. You should use the version switch and search feature of the doc. The docs are changed from version to version.
https://skywalking.apache.org/docs/main/v8.4.0/en/setup/service-agent/java-agent/readme/#the-logic-endpoint

**Comment by lijiawen** [comment]
I don't know what kind of display effect is on the UI of skywalking after adding logic-span, I don't know if there is a sample picture, if it meets our requirements, We will push forward the upgrade of our OAP server, thanks!

![image](https://user-images.githubusercontent.com/3766355/163362239-d491e572-f776-4bb6-87fa-656009ea71fc.png)
maybe like this ,will a new block be added to show the performance metrics of the logic-endpoint separately in dashboard?

**Reply by wu-sheng** [reply]
Logic endpoint would be shown as an endpoint on the top selector.

**Comment by lijiawen** [comment]
![image](https://user-images.githubusercontent.com/3766355/163389068-7135ec76-9a8c-43cd-841e-55262ff3bc41.png)

After I added the login-span tag, I can't seem to find the corresponding trace name in the UI, both dashboard and tracing.

**Comment by lijiawen** [comment]
![image](https://user-images.githubusercontent.com/3766355/163389976-1efa99c0-2962-4f9f-b094-bd6bcf3cb140.png)

**Reply by wu-sheng** [reply]
I think I made a very clear statement, the logic endpoint is for statistics, not for trace queries. Maybe you missed that reply.

**Comment by lijiawen** [comment]
@wu-sheng ，If we want to monitor some metrics of apache-httpclient connection pool, is it possible to do it by skywalking, does this have anything to do with Metrics Exporter?

**Reply by wu-sheng** [reply]
You need to write a plugin like other pool monitoring.
